File: android/app/src/main/java/com/example/demo/domain/DemoViewModel.kt
package com.example.demo.domain

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.demo.data.DemoRepository
import com.example.demo.data.InitializationState
import com.example.demo.data.OpenFeatureInitializer
import com.example.demo.data.PlanConfiguration
import com.example.demo.data.RegionInfo
import com.example.demo.data.RegionPreferences
import com.example.demo.data.SubscriptionPlan
import dagger.hilt.android.lifecycle.HiltViewModel
import dev.openfeature.kotlin.sdk.ImmutableContext
import dev.openfeature.kotlin.sdk.OpenFeatureAPI
import dev.openfeature.kotlin.sdk.Value
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DemoUiState(
    val title: String = "",
    val subscriptionPlans: List<SubscriptionPlan> = emptyList(),
    val isLoading: Boolean = true,
    val initializationError: Boolean = false,
    val currentRegion: RegionInfo? = null,
    val showRegionMenu: Boolean = false,
)

@HiltViewModel
class DemoViewModel
    @Inject
    constructor(
        private val repository: DemoRepository,
        private val openFeatureInitializer: OpenFeatureInitializer,
        private val regionPreferences: RegionPreferences,
    ) : ViewModel() {
        private val _uiState = MutableStateFlow(DemoUiState())
        val uiState: StateFlow<DemoUiState> = _uiState.asStateFlow()

        init {
            loadData()
            observeRegionChanges()
        }

        private fun loadData() {
            viewModelScope.launch {
                // Wait for OpenFeature initialization before loading data
                openFeatureInitializer.initializationState.collect { initState ->
                    when (initState) {
                        InitializationState.LOADING -> {
                            _uiState.value = _uiState.value.copy(isLoading = true, initializationError = false)
                        }
                        InitializationState.SUCCESS -> {
                            loadPlansData()
                        }
                        InitializationState.ERROR -> {
                            _uiState.value =
                                _uiState.value.copy(
                                    isLoading = false,
                                    initializationError = true,
                                )
                        }
                    }
                }
            }
        }

        private fun loadPlansData() {
            viewModelScope.launch {
                combine(
                    repository.getWelcomeMessage(),
                    repository.getPlanConfiguration(),
                ) { title, config ->
                    _uiState.value = _uiState.value.copy(title = title)
                    loadPlans(config)
                }.collect { }
            }
        }

        private fun loadPlans(config: PlanConfiguration) {
            viewModelScope.launch {
                repository.getSubscriptionPlans(config).collect { plans ->
                    _uiState.value =
                        _uiState.value.copy(
                            subscriptionPlans = plans,
                            isLoading = false,
                        )
                }
            }
        }

        private fun observeRegionChanges() {
            viewModelScope.launch {
                regionPreferences.currentRegion.collect { region ->
                    _uiState.value = _uiState.value.copy(isLoading = true)
                    OpenFeatureAPI.setEvaluationContextAndWait(ImmutableContext(attributes = mapOf("region" to Value.String(region.code))))
                    loadPlansData()
                    _uiState.value = _uiState.value.copy(currentRegion = region, isLoading = false)
                }
            }
        }

        fun toggleRegionMenu() {
            _uiState.value = _uiState.value.copy(showRegionMenu = !_uiState.value.showRegionMenu)
        }

        fun selectRegion(regionCode: String?) {
            regionPreferences.setRegionOverride(regionCode)
            _uiState.value = _uiState.value.copy(showRegionMenu = false)
            // Reload data with new region
            if (!_uiState.value.isLoading) {
                _uiState.value = _uiState.value.copy(isLoading = true)
                loadPlansData()
            }
        }

        fun clearRegionOverride() {
            selectRegion(null)
        }
    }

File: ios/ConfidenceDemoApp/ViewModels/DemoViewModel.swift
import Foundation
import Combine
import OpenFeature

struct DemoUiState {
    var title: String = ""
    var subscriptionPlans: [SubscriptionPlan] = []
    var isLoading: Bool = true
    var initializationError: Bool = false
    var currentRegion: RegionInfo?
    var showRegionMenu: Bool = false
}

class DemoViewModel: ObservableObject {
    @Published var uiState = DemoUiState()

    private let repository = DemoRepository()
    private let openFeatureInitializer = OpenFeatureInitializer()
    private let regionPreferences = RegionPreferences()
    private var cancellables = Set<AnyCancellable>()

    init() {
        setupObservers()
        loadData()
    }

    private func setupObservers() {
        // Observe initialization state
        openFeatureInitializer.$initializationState
            .sink { [weak self] initState in
                switch initState {
                case .loading:
                    self?.uiState.isLoading = true
                    self?.uiState.initializationError = false
                case .success:
                    self?.loadPlansData()
                case .error:
                    self?.uiState.isLoading = false
                    self?.uiState.initializationError = true
                }
            }
            .store(in: &cancellables)

        // Observe region changes
        regionPreferences.$currentRegion
            .sink { [weak self] region in
                self?.uiState.currentRegion = region
                if let region = region {
                    self?.setEvaluationContext(region: region.code)
                }
            }
            .store(in: &cancellables)
    }

    private func loadData() {
        // Initial load will be triggered by initialization state observer
    }

    private func loadPlansData() {
        Publishers.CombineLatest(
            repository.getWelcomeMessage(),
            repository.getPlanConfiguration()
        )
        .flatMap { [weak self] title, config -> AnyPublisher<[SubscriptionPlan], Never> in
            self?.uiState.title = title
            return self?.repository.getSubscriptionPlans(config: config) ?? Just([]).eraseToAnyPublisher()
        }
        .receive(on: DispatchQueue.main)
        .sink { [weak self] plans in
            self?.uiState.subscriptionPlans = plans
            self?.uiState.isLoading = false
        }
        .store(in: &cancellables)
    }

    private func setEvaluationContext(region: String) {
        Task {
            let context = ImmutableContext(targetingKey: "user", structure: ImmutableStructure(attributes: ["region": Value.string(region)]))
            await OpenFeatureAPI.shared.setEvaluationContextAndWait(evaluationContext: context)

            DispatchQueue.main.async {
                self.loadPlansData()
            }
        }
    }

    func toggleRegionMenu() {
        uiState.showRegionMenu.toggle()
    }

    func selectRegion(_ regionCode: String?) {
        regionPreferences.setRegionOverride(regionCode)
        uiState.showRegionMenu = false

        // Reload data with new region
        if !uiState.isLoading {
            uiState.isLoading = true
            loadPlansData()
        }
    }

    func clearRegionOverride() {
        selectRegion(nil)
    }
}
